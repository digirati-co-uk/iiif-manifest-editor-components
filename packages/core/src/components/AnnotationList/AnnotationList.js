import React from 'react';
import PropTypes from 'prop-types';
import { Droppable, Draggable } from 'react-beautiful-dnd';
import { IconButton, Typography, withStyles, Divider } from '@material-ui/core';
import { Cancel } from '@material-ui/icons';

import Panel from '../Panel/Panel';
import DefaultAnnotationListToolbar from '../DefaultAnnotationListToolbar/DefaultAnnotationListToolbar';
import LocaleString from '../LocaleString/LocaleString';
import Tooltip from '../DefaultTooltip/DefaultTooltip';
import { EditorConsumer } from '../EditorContext/EditorContext';

const getItemStyle = (isDragging, draggableStyle) => ({
  outline: isDragging ? '2px solid rgb(89, 191, 236)' : '0',
  ...draggableStyle,
});

const emptyFn = () => {};

const style = theme => ({
  list: {
    width: '100%',
    padding: theme.spacing.unit,
    minHeight: '100%',
    position: 'relative',
  },
  annotationDraggable: {
    userSelect: 'none',
    padding: `${theme.spacing.unit / 2}px 0 ${theme.spacing.unit / 2}px ${
      theme.spacing.unit
    }px`,
    margin: `0`,
    background: theme.palette.background.paper,
    display: 'flex',
    flexDirection: 'row',
    alingItems: 'center',
    justifyContent: 'stretch',
    alignItems: 'center',
    borderBottom: `1px solid ${theme.palette.divider}`,
  },
  defaultAnnotation: {
    flex: 1,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'stretch',
    maxWidth: 'calc(100% - 40px)',
  },
  defaultAnnotationText: {
    flex: 1,
    padding: '0 0 0 1rem',
    overflow: 'hidden',
  },
  defaultAnnotationTypo: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
  defaultNoAnnotationIndicator: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

const AnnotationList = ({
  classes,
  children,
  annotations,
  toolbar,
  selected,
  lang,
  select,
  remove,
  invokeAction,
}) => (
  <Panel>
    {toolbar ? (
      toolbar
    ) : (
      <DefaultAnnotationListToolbar invokeAction={invokeAction} />
    )}
    <Divider />
    <Panel.Content>
      <EditorConsumer>
        {configuration => (
          <Droppable droppableId="annotationlist">
            {(providedDroppable, snapshotDroppable) => (
              <div ref={providedDroppable.innerRef} className={classes.list}>
                {annotations ? (
                  annotations.map((annotation, index) => (
                    <Draggable
                      key={annotation.id}
                      draggableId={annotation.id}
                      index={index}
                    >
                      {(provided, snapshot) => (
                        <div
                          ref={provided.innerRef}
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          style={getItemStyle(
                            snapshot.isDragging,
                            provided.draggableProps.style
                          )}
                          className={classes.annotationDraggable}
                        >
                          {typeof children === 'function' ? (
                            children(annotation, remove, select)
                          ) : (
                            <div
                              key={annotation.id}
                              className={classes.defaultAnnotation}
                              onClick={ev => {
                                select(annotation);
                              }}
                            >
                              {configuration.annotation[
                                [
                                  annotation.body.type,
                                  annotation.motivation,
                                ].join('::')
                              ]
                                ? configuration.annotation[
                                    [
                                      annotation.body.type,
                                      annotation.motivation,
                                    ].join('::')
                                  ].icon({
                                    color:
                                      selected === annotation.id
                                        ? 'primary'
                                        : 'inherit',
                                  })
                                : ''}
                              <div
                                className={classes.defaultAnnotationText}
                                title={
                                  annotation.label && annotation.label[lang]
                                    ? annotation.label[lang]
                                    : annotation.id
                                }
                              >
                                <Typography
                                  color={
                                    selected === annotation.id
                                      ? 'primary'
                                      : 'inherit'
                                  }
                                  variant="subtitle2"
                                  className={classes.defaultAnnotationTypo}
                                >
                                  <LocaleString
                                    fallback={annotation.id}
                                    lang={lang}
                                  >
                                    {annotation.label}
                                  </LocaleString>
                                </Typography>
                              </div>
                            </div>
                          )}
                          <Tooltip title="Delete Annotation" placement="right">
                            <IconButton onClick={() => remove(annotation)}>
                              <Cancel />
                            </IconButton>
                          </Tooltip>
                        </div>
                      )}
                    </Draggable>
                  ))
                ) : (
                  <Typography
                    variant="caption"
                    className={classes.defaultNoAnnotationIndicator}
                  >
                    No Annotations
                  </Typography>
                )}
                {providedDroppable.placeholder}
              </div>
            )}
          </Droppable>
        )}
      </EditorConsumer>
    </Panel.Content>
  </Panel>
);

AnnotationList.propTypes = {
  /* JSS classes */
  classes: PropTypes.any,
  /* custom item renderer function fn(annotation:obj, remove: func, select: func)*/
  children: PropTypes.func,
  /* annotations in this specific list */
  annotations: PropTypes.array,
  /* custom toolbar, if annotation type restrictions required, or for creating custom types */
  toolbar: PropTypes.any,
  /* Pa */
  selected: PropTypes.string,
  /* current language */
  lang: PropTypes.string,
  /* on annotation seleced callback */
  select: PropTypes.func,
  /* on remove callback */
  remove: PropTypes.func,
  /* toolbar action dispacher */
  invokeAction: PropTypes.func.isRequired,
};

AnnotationList.defaultProps = {
  selected: null,
  select: emptyFn,
  remove: emptyFn,
  invokeAction: emptyFn,
};

export default withStyles(style)(AnnotationList);
