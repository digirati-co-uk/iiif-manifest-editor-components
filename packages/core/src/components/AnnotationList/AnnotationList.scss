.annotation-list {
  display: flex;
  flex-direction: column;
  &__toolbar {
    height: 60px;
  }
  &__content {
    flex: 1;
  }
}