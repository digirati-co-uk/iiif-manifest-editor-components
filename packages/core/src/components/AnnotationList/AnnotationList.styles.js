const style = theme => ({
  list: {
    width: '100%',
    padding: theme.spacing.unit,
    position: 'relative',
  },
  annotationDraggable: {
    userSelect: 'none',
    padding: `${theme.spacing.unit / 2}px 0 ${theme.spacing.unit / 2}px ${
      theme.spacing.unit
    }px`,
    margin: `0`,
    background: theme.palette.background.paper,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'stretch',
    alignItems: 'center',
    borderBottom: `1px solid ${theme.palette.divider}`,
  },
  annotationDragging: {
    outline: `2px solid ${theme.palette.primary.main}`,
  },
  defaultNoAnnotationIndicator: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default style;
