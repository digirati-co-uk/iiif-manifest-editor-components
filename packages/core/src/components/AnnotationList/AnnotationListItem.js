import React from 'react';
import { Typography, withStyles } from '@material-ui/core';
import LocaleString from '../LocaleString/LocaleString';
import { EditorConsumer } from '../EditorContext/EditorContext';

const style = theme => ({
  main: {
    flex: 1,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'stretch',
    maxWidth: 'calc(100% - 40px)',
  },
  textBlock: {
    flex: 1,
    padding: '0 0 0 1rem',
    overflow: 'hidden',
  },
  typo: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
});

const AnnotationListItem = ({
  classes,
  annotation,
  lang,
  onSelect,
  isSelected,
}) => {
  const internalAnnotationType = [
    annotation.body.type,
    annotation.motivation,
  ].join('::');
  const textColor = isSelected ? 'primary' : 'inherit';
  return (
    <EditorConsumer>
      {configuration => (
        <div
          key={annotation.id}
          className={classes.main}
          onClick={() => onSelect(annotation)}
        >
          {configuration.annotation[internalAnnotationType]
            ? configuration.annotation[internalAnnotationType].icon({
                color: textColor,
              })
            : ''}
          <div
            className={classes.textBlock}
            title={
              annotation.label && annotation.label[lang]
                ? annotation.label[lang]
                : annotation.id
            }
          >
            <Typography
              color={textColor}
              variant="subtitle2"
              className={classes.typo}
            >
              <LocaleString fallback={annotation.id} lang={lang}>
                {annotation.label}
              </LocaleString>
            </Typography>
          </div>
        </div>
      )}
    </EditorConsumer>
  );
};
export default withStyles(style)(AnnotationListItem);
