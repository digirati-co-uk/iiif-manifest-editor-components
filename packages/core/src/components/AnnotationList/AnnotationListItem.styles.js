export default theme => ({
  main: {
    flex: 1,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'stretch',
    maxWidth: 'calc(100% - 40px)',
  },
  textBlock: {
    flex: 1,
    padding: '0 0 0 1rem',
    overflow: 'hidden',
  },
  label: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
});
