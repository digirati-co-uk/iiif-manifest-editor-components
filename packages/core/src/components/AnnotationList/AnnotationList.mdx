---
name: Annotation List
menu: 02. High Level Components
---

import { Playground, PropsTable } from 'docz';
import AnnotationList from './AnnotationList'; 
import ManifestEditor from '../ManifestEditor/ManifestEditor.js';

# Annotation List

Component that lists annotations from a specific path.

<PropsTable of={AnnotationList} />

## Demo as an individual Component

<Playground>
  <ManifestEditor>
  <div 
    style={{
      display: 'flex',
      alignItems: 'stretch',   
      justifyContent: 'stretch',
      width: 250,
      height: 100,
      height: 400,
    }}
  >
    <AnnotationList
      annotations={[
        {
          id: 'test-1',
          body: {
            type: 'Image',
          }
        },
        {
          id: 'test-2',
          body: {
            type: 'Video',
          }
        },
        {
          id: 'test-3',
          body: {
            type: 'Audio',
          }
        },
        {
          id: 'test-4',
          body: {
            type: 'TextualBody',
          }
        },
        {
          id: 'test-5',
          body: {
            type: 'Image',
          }
        }
        ,
        {
          id: 'test-6',
          body: {
            type: 'Video',
          }
        }
      ]}
    />
  </div>
  </ManifestEditor>
</Playground>


## TODO: Two annotation lists linking Drag and drop tests.

<Playground>
  <ManifestEditor>
  <div 
    style={{
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'stretch',   
      justifyContent: 'stretch',
      width: 500,
      height: 100,
      height: 400,
    }}
  >
    <div style={{ width: 250 }}>
      <AnnotationList
        source='items'
        annotations={[
          {
            id: 'test-1',
            body: {
              type: 'Image',
            }
          },
          {
            id: 'test-2',
            body: {
              type: 'TextualBody',
            }
          }
        ]}
      />
    </div>
    <div style={{ width: 250 }}>
      <AnnotationList
        source='annotations'
        annotations={[
          {
            id: 'test-3',
            body: {
              type: 'Video',
            }
          },
          {
            id: 'test-4',
            body: {
              type: 'Image',
            }
          }
        ]}
      />
    </div>
  </div>
  </ManifestEditor>
</Playground>


## Override the default annotation renderer

<Playground>
  <div 
    style={{
      display: 'flex',
      alignItems: 'stretch',   
      justifyContent: 'stretch',
      width: 250,
      height: 100,
      height: 400,
    }}
  >
    <ManifestEditor>
      <AnnotationList
        annotations={[
          {
            id: 'test-1',
            body: {
              type: 'Video',
            }
          },
          {
            id: 'test-2',
            body: {
              type: 'Video',
            }
          }
        ]}
      >
        {(annotation, deleteAction, toggleSelectAction)=> (
          <div>Custom annotation renderer {annotation.id}</div>
        )}
      </AnnotationList>
    </ManifestEditor>
  </div>
</Playground>


## Requirements

- [x] add drag and drop list using the same drag drop manager as all the other components
- [x] add selectable state for annotations
- [x] use fesk-bem for themeability
- [x] use the global selected resources
- [x] make sure that it is able to work as an idividual component
- [ ] being able to specify that where the displayed annotations coming form [items/annotations]
- [x] being able to override the default annotation renderer
- [x] create default toolbar component
- [x] add delete button for the individual annotations
