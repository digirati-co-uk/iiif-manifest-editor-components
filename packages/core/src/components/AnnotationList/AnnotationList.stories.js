import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import AnnotationList from './AnnotationList.js';
import AnnotationListItem from './AnnotationListItem';
import { DragDropContext } from 'react-beautiful-dnd';

storiesOf('AnnotationList', module)
  .addDecorator(withKnobs)
  .add('AnnotationList', () => {
    const getResource = id => ({
      id,
      label: { en: [id] },
      type: 'Annotation',
      motivation: 'describing',
      body: {
        type: 'TextualBody',
      },
    });
    return (
      <DragDropContext onDragEnd={() => {}}>
        <AnnotationList
          annotations={['test-annotation-1', 'test-annotation-2']}
          getResource={getResource}
          selected={text('Selected', null)}
          selectedColor={text('Selected Color', 'primary')}
          isEditingAllowed={boolean('Is Editing Allowed', true)}
        />
      </DragDropContext>
    );
  })
  .add('AnnotationListItem', () => {
    const annotation = {
      body: text('Annotation Body', 'Some content here'),
      id: text('Annotation id', 'Some default id'),
      label: text('Label', 'Annotation Label text'),
      classes: null,
      onSelect: () => {},
    };
    return (
      <DragDropContext onDragEnd={() => {}}>
        <AnnotationListItem
          selectedColor={text('Selected solor', 'primary')}
          lang={'en-gb'}
          isSelected={boolean('Is selected', true)}
          annotation={annotation}
        />
      </DragDropContext>
    );
  });
