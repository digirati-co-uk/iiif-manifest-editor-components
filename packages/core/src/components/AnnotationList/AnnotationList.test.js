import * as React from 'react';
import AnnotationList from './AnnotationList';

describe('AnnotationList', () => {
  it('renders the annotations list unchanged', () => {
    const TEST_ANNOTATIONS = [
      {
        id: 'test-annotation-1',
      },
      {
        id: 'test-annotation-2',
      },
    ];

    const annotationList = shallow(
      <AnnotationList annotations={TEST_ANNOTATIONS} />
    );
    expect(annotationList).toMatchSnapshot();
  });
});
