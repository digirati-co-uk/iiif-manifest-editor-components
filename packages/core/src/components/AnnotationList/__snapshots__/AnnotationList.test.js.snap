// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationList renders the annotations list unchanged 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(AnnotationList)
    annotations={
      Array [
        Object {
          "id": "test-annotation-1",
        },
        Object {
          "id": "test-annotation-2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "annotations": Array [
        Object {
          "id": "test-annotation-1",
        },
        Object {
          "id": "test-annotation-2",
        },
      ],
      "classes": Object {
        "annotationDraggable": "AnnotationList-annotationDraggable-2",
        "defaultAnnotation": "AnnotationList-defaultAnnotation-3",
        "defaultAnnotationText": "AnnotationList-defaultAnnotationText-4",
        "defaultAnnotationTypo": "AnnotationList-defaultAnnotationTypo-5",
        "defaultNoAnnotationIndicator": "AnnotationList-defaultNoAnnotationIndicator-6",
        "list": "AnnotationList-list-1",
      },
      "iconProvider": [Function],
      "invokeAction": [Function],
      "remove": [Function],
      "select": [Function],
      "selected": null,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "annotations": Array [
          Object {
            "id": "test-annotation-1",
          },
          Object {
            "id": "test-annotation-2",
          },
        ],
        "classes": Object {
          "annotationDraggable": "AnnotationList-annotationDraggable-2",
          "defaultAnnotation": "AnnotationList-defaultAnnotation-3",
          "defaultAnnotationText": "AnnotationList-defaultAnnotationText-4",
          "defaultAnnotationTypo": "AnnotationList-defaultAnnotationTypo-5",
          "defaultNoAnnotationIndicator": "AnnotationList-defaultNoAnnotationIndicator-6",
          "list": "AnnotationList-list-1",
        },
        "iconProvider": [Function],
        "invokeAction": [Function],
        "remove": [Function],
        "select": [Function],
        "selected": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
